File: classes04/python/webcamHeadPose.py
import sys
import cv2
import dlib
import numpy as np
from renderFace import renderFace

SKIP_FRAMES = 5
RESIZE_HEIGHT = 320
PREDICTOR_PATH = "../../common/resources/zxm_shape_predictor_70_face_landmarks.dat"

def get3dModelPoints():
    """
    获取3D模型上的点
    Returns:
        np.array:3D模型点的数组
    """
    pass
    modelPoints = [[0.0, 0.0, 0.0],
                   [0.0, -330.0, -65.0],
                   [-225.0, 170.0, -135.0],
                   [225.0, 170.0, -135.0],
                   [-150.0, -150.0, -125.0],
                   [150.0, -150.0, -125.0]]
    return np.array(modelPoints, dtype=np.float64)

def get2dImagePoints(shape):
    """
    获取2D图像上的标记点
    Args:
        full_object_detection:dlib的识别结果
    Returns:
        np.array:2D图像上的标志点
    """
    pass
    imagePoints = [[shape.part(30).x, shape.part(30).y],
                   [shape.part(8).x, shape.part(8).y],
                   [shape.part(36).x, shape.part(36).y],
                   [shape.part(45).x, shape.part(45).y],
                   [shape.part(48).x, shape.part(48).y],
                   [shape.part(54).x, shape.part(54).y]]
    
    return np.array(imagePoints, dtype=np.float64)

def getCameraMatrix(flocalLength, center):
    """
    获取摄像头的变换矩阵
    Args:
        flocalLength:(float),摄像头的焦距;
        center:[x,y],焦点
    Returns:
        np.array
    """
    pass

    cameraMatrix = [[flocalLength, 0, center[0]],
                    [0, flocalLength, center[1]],
                    [0, 0, 1]]
    return np.array(cameraMatrix, dtype=np.float64)


if __name__=="__main__":
    try:
        capture = cv2.VideoCapture(0)
        if (capture.isOpened() is False):
            print("不能打开摄像头!")
            sys.exit(0)
        
        fps = 30.0
        ret, im = capture.read()
        if ret == True:
            height = im.shape[0]
            RESIZE_SCALE = float(height)/RESIZE_HEIGHT
            size = im.shape[0:2]
        else:
            print("不能读取视频中的内容")
            sys.exit(0)
        
        detector = dlib.get_frontal_face_detector()
        predictor = dlib.shape_predictor(PREDICTOR_PATH)

        t = cv2.getTickCount()
        count = 0

        while(True):
            if count ==0:
                t=cv2.getTickCount()
            
            ret, im=capture.read()
            imSmall = cv2.resize(im, None, fx=1.0/RESIZE_SCALE, fy=1.0/RESIZE_SCALE, interpolation=cv2.INTER_LINEAR)

            if(count % SKIP_FRAMES ==0):
                faces = detector(imSmall, 0)
                modelPoints = get3dModelPoints()

                for face in faces:
                    newRect = dlib.rectangle(int(face.left()*RESIZE_SCALE),
                                             int(face.top()*RESIZE_SCALE),
                                             int(face.right()*RESIZE_SCALE),
                                             int(face.bottom()*RESIZE_SCALE))
                    shape = predictor(im, newRect)
                    # renderFace(im, shape)

                    iamgePoints = get2dImagePoints(shape)

                    rows, cols, ch = im.shape
                    flocalLength = cols
                    cameraMatrix = getCameraMatrix(flocalLength, (rows/2, cols/2))

                    distCoeffs = np.zeros((4,1), dtype=np.float64)

                    success, rotationVector, translationVector = cv2.solvePnP(modelPoints, iamgePoints, cameraMatrix, distCoeffs)

                    noseEndPoints3D = np.array([[0, 0, 1000.0]], dtype=np.float64)
                    noseEndPoint2D, jacobain = cv2.projectPoints(noseEndPoints3D, rotationVector, translationVector, cameraMatrix, distCoeffs)

                    print("noseEndPoint2D:\n{}".format(noseEndPoint2D))
                    print("jacobain:\n{}".format(jacobain))

                    p1 = (int(iamgePoints[0,0]), int(iamgePoints[0,1]))
                    p2 = (int(noseEndPoint2D[0,0,0]), int(noseEndPoint2D[0,0,1]))

                    cv2.line(im, p1, p2, (100, 220, 2), thickness=2)

                
                
                cv2.imshow("webcam head pose", im)

                
            
            if(cv2.waitKey(1)&0xFF)==27:
                sys.exit()
            
            count = count+1

    except Exception as e:
        print(e)
    
    
    

File: classes04/cpp/webcamHeadPose.cpp
#include <dlib/opencv.h>
#include <opencv2/opencv.hpp>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing/render_face_detections.h>
#include <dlib/image_processing.h>
#include <dlib/gui_widgets.h>
#include "../../classes03/cpp/renderFace.hpp"

using namespace dlib;
using namespace std;
using namespace cv;

#define FACE_DOWNSAMPLE_RATIO 2
#define SKIP_FRAMES 10
#define OPENCV_FACE_RENDER

/**
 * @brief 获取参考点的3D坐标(U,V,W)
 * 
 * @return std::vector<cv::Point3d> 
 */
std::vector<cv::Point3d> get3dModelPoints()
{
    std::vector<cv::Point3d> modelPoints;
    modelPoints.push_back(cv::Point3d(0, 0, 0));   //使用OpenCV的POSIT时，第一个点必须是(0,0,0)
    modelPoints.push_back(cv::Point3d(0, -330, -65));
    modelPoints.push_back(cv::Point3d(-225, 170, -135));
    modelPoints.push_back(cv::Point3d(225, 170, -135));
    modelPoints.push_back(cv::Point3d(-150, -150, -125));
    modelPoints.push_back(cv::Point3d(150, -150, -125));

    return modelPoints;
}

/**
 * @brief 获取图片上的2D参照点
 * 
 * @param d full_object_detection 对象检测
 * @return std::vector<cv::Point2d> 
 */
std::vector<cv::Point2d> get2dImagePoints(dlib::full_object_detection &d)
{
    std::vector<cv::Point2d> imagePoints;
    imagePoints.push_back( cv::Point2d(d.part(30).x(), d.part(30).y() ));        // 鼻尖点
    imagePoints.push_back( cv::Point2d(d.part(8).x(), d.part(8).y() ));          // 下巴
    imagePoints.push_back( cv::Point2d(d.part(36).x(), d.part(36).y() ));        // 左眼角点
    imagePoints.push_back( cv::Point2d(d.part(45).x(), d.part(45).y() ));        // 右眼角点
    imagePoints.push_back( cv::Point2d(d.part(48).x(), d.part(48).y() ));        // 左嘴角点
    imagePoints.push_back( cv::Point2d(d.part(54).x(), d.part(54).y() ));        // 右嘴角点

    return imagePoints;
}

/**
 * @brief 获取摄像头的矩阵
 * 
 * @param focal_length 矩阵
 * @param center 图片中心点
 * @return cv::Mat 返回的3x3的矩阵
 */
cv::Mat getCameraMatrix(float focal_length, cv::Point2d center)
{
    cv::Mat cameraMatrix = (cv::Mat_<double>(3,3) << focal_length, 0 , center.x, 0, focal_length, center.y, 0, 0, 1);
    return cameraMatrix;
}

int main(int argc, char const *argv[])
{
    try
    {
        cv::VideoCapture capture(0);
        if(!capture.isOpened())
        {
            cerr<< "不能打开摄像头"<<endl;
            return EXIT_FAILURE;
        }

        double fps = 30.0f;
        cv::Mat im;

        capture >> im;
        cv::Mat imSmall, imDisplay;

        cv::resize(im, imSmall, cv::Size(), 1.0/FACE_DOWNSAMPLE_RATIO, 1.0/FACE_DOWNSAMPLE_RATIO);
        cv::resize(im, imDisplay, cv::Size(), 0.5, 0.5);

        cv::Size size = im.size();

        dlib::frontal_face_detector detector = dlib::get_frontal_face_detector();
        dlib::shape_predictor predictor;

        dlib::deserialize("../../common/resources/zxm_shape_predictor_70_face_landmarks.dat") >> predictor;

        int count = 0;
        double t = (double)cv::getTickCount();

        std::vector<dlib::rectangle> faces;

        while(1)
        {

            if(count == 0)
                t = cv::getTickCount();
            
            capture >> im;

            cv::resize(im, imSmall, cv::Size(), 1.0/FACE_DOWNSAMPLE_RATIO, 1.0/FACE_DOWNSAMPLE_RATIO);

            cv_image<bgr_pixel> cimgSmall(imSmall);
            cv_image<bgr_pixel> cimg(im);

            if(count % SKIP_FRAMES == 0)
            {
                faces = detector(cimgSmall);
            }

            std::vector<cv::Point3d> modelPoints = get3dModelPoints();
            std::vector<dlib::full_object_detection> shapes;

            for(unsigned long i = 0; i<faces.size(); i++)
            {
                dlib::rectangle r( (long)(faces[i].left()*FACE_DOWNSAMPLE_RATIO),
                                   (long)(faces[i].top()*FACE_DOWNSAMPLE_RATIO),
                                   (long)(faces[i].right()*FACE_DOWNSAMPLE_RATIO),
                                   (long)(faces[i].bottom()*FACE_DOWNSAMPLE_RATIO)
                                 );
                
                dlib::full_object_detection shape = predictor(cimg, r);
                shapes.push_back(shape);

                renderFace(im, shape);

                std::vector<cv::Point2d> imagePoints  = get2dImagePoints(shape);

                // 摄像头参数
                double focal_length = im.cols;
                cv::Mat cameraMatrix = getCameraMatrix(focal_length, cv::Point2d(im.cols/2, im.rows/2));

                // 假设摄像头没有径向畸变
                cv::Mat distCoeffs = cv::Mat::zeros(4,1,cv::DataType<double>::type);

                // 使用solvePnP计算rotation和translation
                cv::Mat rotationVector;
                cv::Mat translationVector;
                cv::solvePnP(modelPoints, imagePoints, cameraMatrix, distCoeffs, rotationVector, translationVector);

                std::vector<cv::Point3d> noseEndPoint3D;
                std::vector<cv::Point2d> noseEndPoint2D;
                noseEndPoint3D.push_back(cv::Point3d(0,0,1000));
                cv::projectPoints(noseEndPoint3D, rotationVector, translationVector, cameraMatrix, distCoeffs, noseEndPoint2D);

                cv::line(im, imagePoints[0], noseEndPoint2D[0], cv::Scalar(255,0,255), 2);
            }
            
            

            cv::putText(im, cv::format("fps %.2f", fps), cv::Point(50, size.height-50), cv::FONT_HERSHEY_COMPLEX, 1.0, cv::Scalar(0, 255, 255));

            imDisplay = im;
            cv::resize(im, imDisplay, cv::Size(), 0.5, 0.5);
            cv::imshow("Webcam Head Pose", imDisplay);

            if(count % 15 == 0)
            {
                int k = cv::waitKey(1);
                if(k == 'q' || k==27)
                {
                    break;
                }
            }

            count++;
            if(count == 100)
            {
                t=((double)cv::getTickCount()-t)/cv::getTickFrequency();
                fps = 100.0/t;
                count = 0;
            }
        }
    }
    catch(serialization_error& e)
    {
        cout << "识别模型没有找到"<<endl;
        cout << endl << e.what() << endl;
    }
    catch(exception& e)
    {
        cout << e.what() <<endl;
    }
    return 0;
}

